### ISSUE TITLE
1.20.2 -> 1.21 broke my deployment

### ISSUE THREAD

**Issue by bebbi** [serverless/serverless#4182]
# This is a (Bug Report / Feature Proposal)

Bug

## Description

After updating various deps I have strange `Data error`s after deploying the serverless project which I've isolated to serverless upgrade from 1.20.2 to the latest release 1.21.0

For bug reports:
* What went wrong?

Deployment worked without error but calls to the lambdas result in strange errors, see example traces.

* What did you expect should have happened?

Same as 1.20.2, just works.
I've gone back and forth with different configurations for other deps and can confirm 100% success with 1.20.2 and 100% failure on 1.21

* What was the config you used?

excerpts - let me know if more needed

    plugins:
      - serverless-webpack
    ...
    functions:
      myfunc:
        handler: handler.myHandler
        events:
          - http:
              path: myhandler
              method: post
              cors: true
              integration: lambda
              request:
                passThrough: NEVER
              response:
                headers:
                  Content-Type: "'application/json'"
        name: ${self:provider.stage}-myfunc
        description: 'some'
        environment: <some env>

* What stacktrace or error message from your provider did you see?

For feature proposals:
* What is the use case that should be solved. The more detail you describe this in the easier it is to understand for us.
* If there is additional config how would it look

Similar or dependent issues:

## Additional Data

* ***Serverless Framework Version you're using***:
See title
* ***Operating System***:
osx
* ***Stack Trace***:

Ex 1

    {
      "errorMessage": "Data error",
      "errorType": "Error",
      "stackTrace": [
        "Object.Module._extensions..js (module.js:579:10)",
        "Module.load (module.js:487:32)",
        "tryModuleLoad (module.js:446:12)",
        "Function.Module._load (module.js:438:3)",
        "Module.require (module.js:497:17)",
        "require (internal/module.js:20:19)",
        "Object.<anonymous> (/var/task/node_modules/pdfmake/src/textTools.js:4:19)",
        "Module._compile (module.js:570:32)",
        "Object.Module._extensions..js (module.js:579:10)",
        "Module.load (module.js:487:32)",
        "tryModuleLoad (module.js:446:12)"
      ]
    }

Ex 2

    {
      "errorMessage": "Data error",
      "errorType": "Error",
      "stackTrace": [
        "new UnicodeTrie (/var/task/node_modules/unicode-trie/index.js:57:14)",
        "Object.<anonymous> (/var/task/node_modules/linebreak/src/linebreaker.js:17:15)",
        "Object.<anonymous> (/var/task/node_modules/linebreak/src/linebreaker.js:161:4)",
        "Module._compile (module.js:570:32)",
        "Object.Module._extensions..js (module.js:579:10)",
        "Module.load (module.js:487:32)",
        "tryModuleLoad (module.js:446:12)",
        "Function.Module._load (module.js:438:3)",
        "Module.require (module.js:497:17)",
        "require (internal/module.js:20:19)",
        "Object.<anonymous> (/var/task/node_modules/pdfmake/src/textTools.js:4:19)",
        "Module._compile (module.js:570:32)",
        "Object.Module._extensions..js (module.js:579:10)",
        "Module.load (module.js:487:32)",
        "tryModuleLoad (module.js:446:12)"
      ]
    }

* ***Provider Error messages***:

**Comment by HyperBrain** [comment]
From the callstacks it seems that some dependencies were not packaged and when the deployed Lambda tries to load them it fails. Can you do a `serverless package` and check the contents of the zip files in `.serverless` afterwards? There should be something missing in node_modules.

**Comment by jaikdean** [comment]
I'm seeing the same problem. I have `"deasync": "^0.1.10"` in my dependencies. It's being packaged correctly, but I get the following error when the Lambda function runs:

```
module initialization error: Error
→ /var/task/node_modules/deasync/Debug/deasync.node
→ /var/task/node_modules/deasync/out/Release/deasync.node
→ /var/task/node_modules/deasync/Release/deasync.node
→ /var/task/node_modules/deasync/build/default/deasync.node
→ /var/task/node_modules/deasync/compiled/6.10.3/linux/x64/deasync.node
at bindings (/var/task/node_modules/bindings/bindings.js:88:9)
at /var/task/node_modules/deasync/index.js:34:32
at Object.<anonymous> (/var/task/node_modules/deasync/index.js:76:2)
at Module._compile (module.js:570:32)
at Object.Module._extensions..js (module.js:579:10)
at Module.load (module.js:487:32)
at tryModuleLoad (module.js:446:12)
at Function.Module._load (module.js:438:3)
at Module.require (module.js:497:17)
at require (internal/module.js:20:19)
```

**Comment by jaikdean** [comment]
It looks like binary files are being corrupted when they are zipped, looking at the files packaged by 1.21.0 vs 1.20.2, the file size differs. This seems to be the case with all binary files I've found:

```bash
$ ls -al 1.20.2/node_modules/deasync/bin/linux-x64-node-6/deasync.node 
-rwxr-xr-x  1 jaikdean  staff  13989 31 Dec  1979 1.20.2/node_modules/deasync/bin/linux-x64-node-6/deasync.node
$ ls -al 1.21.0/node_modules/deasync/bin/linux-x64-node-6/deasync.node 
-rwxr-xr-x  1 jaikdean  staff  15613 31 Dec  1979 1.21.0/node_modules/deasync/bin/linux-x64-node-6/deasync.node```

**Comment by jaikdean** [comment]
![screen shot 2017-08-31 at 14 44 56](https://user-images.githubusercontent.com/348408/29926335-0b933e00-8e5b-11e7-8866-c5fe1f360ba5.png)
1.20.2 on the left, 1.21.0 on the right

**Comment by cadams500** [comment]
We are having this issue as well. All of our binary Java classfiles are being corrupted and modified when packaging.

**Comment by RafalWilinski** [comment]
I've tried reproducing issue by installing `deasync` module and I was able to reproduce it using Node `8.2.1` on OS X.

At the beginning I thought that issue was related to this PR: https://github.com/serverless/serverless/pull/4145 because it changes the way things are packaged a bit but after reverting these commits everything is still the same. 

Then this PR came to my attention: https://github.com/serverless/serverless/pull/3924/commits so I've run a quick research:
```bash
mkdir ~/Desktop/before
mkdir ~/Desktop/after

# Before PR
git checkout f0b1474f1e49d69c39e27e3c8def929569e20bed
cd ../sls-tests
../serverless/bin/serverless package
cp -R .serverless ~/Desktop/before
cd ../serverless

# After PR
git checkout f4ee5481c6da0d61168bf1b2770958af6d571300
cd ../sls-tests
../serverless/bin/serverless package
cp -R .serverless ~/Desktop/after

# Compare example exec
cd ~/Desktop/before/.serverless
tar -xvf aws-nodejs-tests.zip
cp node_modules/deasync/bin/linux-x64-node-6/deasync.node ~/Desktop/deasync.node.before

cd ~/Desktop/after/.serverless
tar -xvf aws-nodejs-tests.zip
cp node_modules/deasync/bin/linux-x64-node-6/deasync.node ~/Desktop/deasync.node.after

cd ~/Desktop
ls -la | grep deasync
```
Output:
```
-rwxr-xr-x    1 Rafal  staff    15679 Aug 31 21:30 deasync.node.after
-rwxr-xr-x    1 Rafal  staff    13989 Aug 31 21:30 deasync.node.before
```

Bingo. Looking for root cause in code now...

**Comment by cadams500** [comment]
This call, and the encoding of UTF8 (instead of a raw buffer), seems potentially troublesome:
https://github.com/serverless/serverless/pull/3924/files#diff-0e8817ce3b2d950d2b8d0e47f9fccb2eR107

**Comment by RafalWilinski** [comment]
@cadams500 I've investigated that and it doesn't matter at all, at least from my experience.

**Comment by RafalWilinski** [comment]
@cadams500 I was wrong and you are right, the default encoding for `readFileSync` is `null` and here `utf8` is used which causes problems. Thanks! I'll submit a PR in a moment.

**Comment by cadams500** [comment]
@RafalWilinski No worries at all. Thanks for digging and fixing so quickly!

**Comment by pmuens** [comment]
Thanks for reporting @bebbi and thanks everyone for the valuables discussion here.

Sorry for the inconveniences! We'll review #4189 ASAP (thanks @RafalWilinski for fixing this! 🎉).

**Comment by pmuens** [comment]
Hey everyone 👋

We've just published the patch release `v1.21.1` which will include this fix (see: https://github.com/serverless/serverless/releases/tag/v1.21.1).
